src/main/java/org/generationcp/commons/ruleengine/resolver/LocationAbbreviationResolver.java

```diff
@@ -0,0 +1,75 @@
+package org.generationcp.commons.ruleengine.resolver;
+
+import org.apache.commons.lang3.StringUtils;
+import org.generationcp.middleware.domain.etl.MeasurementVariable;
+import org.generationcp.middleware.domain.oms.TermId;
+import org.generationcp.middleware.service.api.dataset.ObservationUnitData;
+import org.generationcp.middleware.service.api.dataset.ObservationUnitRow;
+import org.generationcp.middleware.service.impl.study.StudyInstance;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.CollectionUtils;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Revolves Location value for Nurseries and Trials.
+ */
+public class LocationAbbreviationResolver implements KeyComponentValueResolver {
+
+	protected List<MeasurementVariable> conditions;
+	protected ObservationUnitRow observationUnitRow;
+	protected Map<Integer, StudyInstance> studyInstanceMap;
+
+	private static final Logger LOG = LoggerFactory.getLogger(LocationAbbreviationResolver.class);
+
+	public LocationAbbreviationResolver(final ObservationUnitRow observationUnitRow, final Map<Integer, StudyInstance> studyInstanceMap) {
+
+		this.observationUnitRow = observationUnitRow;
+		this.studyInstanceMap = studyInstanceMap;
+	}
+
+	@Override
+	public String resolve() {
+		String location = "";
+
+		if (!CollectionUtils.isEmpty(studyInstanceMap)) {
+			if (this.observationUnitRow != null) {
+
+				final Optional<ObservationUnitData> instanceNoUnitData =
+					Optional.of(observationUnitRow.getVariables().entrySet().stream().collect(Collectors
+						.toMap(k -> k.getValue().getVariableId(),
+							k -> k.getValue())).get(TermId.TRIAL_INSTANCE_FACTOR.getId()));
+
+				if (instanceNoUnitData.isPresent()) {
```

Comment:
Sonar says `instanceNoUnitData` is always present here, probably because `Optional.of` instead of `Optional.ofNullable`